### PR TITLE
Feature/loss

### PR DESCRIPTION
**Pull request by ZizhenWang**
#298

### CODE REVIEW

File: matchzoo/losses.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    """
+    Support user-defined margin value.
+
+    :param y_true: Label.
+    :param y_pred: Predict result.
+    :return: Hinge loss computed by user-defined margin.
+    """
+    y_pos = Lambda(lambda a: a[::2, :], output_shape=(1,))(y_pred)
+    y_neg_list = []
+    for i in range(_neg_num):
```
Comment (bwanglzu) [inline matchzoo/losses.py]: I don't like the `i` here, can we do something like:
```python
for idx in range(num_neg):
    # do something
```

File: matchzoo/losses.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+def rank_hinge_loss(y_true, y_pred):
+    """
+    Support user-defined margin value.
+
+    :param y_true: Label.
+    :param y_pred: Predict result.
+    :return: Hinge loss computed by user-defined margin.
+    """
+    y_pos = Lambda(lambda a: a[::2, :], output_shape=(1,))(y_pred)
+    y_neg_list = []
```
Comment (bwanglzu) [inline matchzoo/losses.py]: use the same naming conversion:

```python
y_pos = ...
y_neg = []
```

File: matchzoo/losses.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    """
+    Support user-defined negative sample number.
+
+    :param y_true: Label.
+    :param y_pred: Predict result.
+    :return: Crossentropy loss computed by user-defined negative number.
+    """
+    y_pos_logits = Lambda(lambda a: a[::(_neg_num+1), :])(y_pred)
+    y_pos_labels = Lambda(lambda a: a[::(_neg_num+1), :])(y_true)
+    logits_list, labels_list = [y_pos_logits], [y_pos_labels]
```
Comment (bwanglzu) [inline matchzoo/losses.py]:
```python
pos_logits, pos_labels = ...
```
or in line 35-36
```python
pos_logits = list(...)
pos_labels = list(...)
```

File: matchzoo/losses.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    for i in range(_neg_num):
```
Comment (bwanglzu) [inline matchzoo/losses.py]: same, change `i` to a readable variable name.

File: matchzoo/losses.py (hunk 3 above)
Comment (ZizhenWang) [reply matchzoo/losses.py]: write these code in line is too long and not clean, like:
    pos_logits, pos_labels = Lambda(lambda a:
                                                            a[::(_neg_num+1), :])(y_pred),\
                                             Lambda(lambda a:
                                                            a[::(_neg_num+1), :])(y_true)
    logits_list, labels_list = [pos_logits], [pos_labels]

File: matchzoo/losses.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+from keras import backend as K
+from keras import layers
+
+# TODO: Loss function parameters setting.
+margin = 1.0
+neg_num = 1
+
+
+def rank_hinge_loss(y_true: typing.Any,
+                    y_pred: typing.Any) -> float:
```
Comment (pl8787) [inline matchzoo/losses.py]: I think the output value is not type `float`, but type `Tensor`.

File: matchzoo/losses.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    for neg_idx in range(neg_num):
+        y_neg.append(layers.Lambda(lambda a: a[(neg_idx+1)::(neg_num+1), :],
+                                   output_shape=(1,))(y_pred))
+    y_neg = K.max(K.concatenate(y_neg, axis=-1), axis=-1, keepdims=True)
+    loss = K.maximum(0., margin + y_neg - y_pos)
+    return K.mean(loss)
+
+
+def rank_crossentropy_loss(y_true: typing.Any,
+                           y_pred: typing.Any) -> float:
```
Comment (pl8787) [inline matchzoo/losses.py]: The same

File: matchzoo/losses.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
+"""Self defined loss function."""
+import typing
+from keras import backend as K
+from keras import layers
+
+# TODO: Loss function parameters setting.
+margin = 1.0
+neg_num = 1
```
Comment (pl8787) [inline matchzoo/losses.py]: The loss parameter here? Why?

Comment (ZizhenWang) [reply matchzoo/losses.py]: It is concise to set loss parameter as global parameter in losses package, although it may leads to some permission problems.

File: matchzoo/losses.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+
+def rank_hinge_loss(y_true: typing.Any,
+                    y_pred: typing.Any) -> typing.Any:
```
Comment (faneshion) [inline matchzoo/losses.py]: `typing.Any` is not the best choice for the type of parameters. At least, you should list some possible types, such as `np.ndarray`.

Comment (ZizhenWang) [reply matchzoo/losses.py]: The inputs and outputs type of loss function is `Tensor`, which should be represented as `typing.Any`

Comment (bwanglzu) [reply matchzoo/losses.py]: maybe `Typing.Iterable` is better?
